test/retry_test.go

```diff
@@ -550,3 +554,166 @@ func (s) TestRetryStreaming(t *testing.T) {
+				"grpc-status", "0", // OK
+			}},
+		}},
+		refuseStream: func(i uint32) bool {
+			return i == 1
+		},
+	}
+	server.start(t, lis)
+	handler := &retryStatsHandler{}
+	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithInsecure(), grpc.WithStatsHandler(handler),
+		grpc.WithDefaultServiceConfig((`{
+    "methodConfig": [{
+      "name": [{"service": "grpc.testing.TestService"}],
+      "retryPolicy": {
+          "MaxAttempts": 4,
+          "InitialBackoff": ".01s",
+          "MaxBackoff": ".01s",
+          "BackoffMultiplier": 1.0,
+          "RetryableStatusCodes": [ "UNAVAILABLE" ]
+      }
+    }]}`)))
+	if err != nil {
+		t.Fatalf("failed to dial due to err: %v", err)
+	}
+	defer cc.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	client := testpb.NewTestServiceClient(cc)
+
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+		t.Fatalf("unexpected EmptyCall error: %v", err)
+	}
+	handler.mu.Lock()
+	want := []stats.RPCStats{
+		&stats.Begin{},
+		&stats.OutHeader{FullMethod: "/grpc.testing.TestService/EmptyCall"},
+		&stats.OutPayload{WireLength: 5},
+		&stats.End{},
+
+		&stats.Begin{IsTransparentRetryAttempt: true},
+		&stats.OutHeader{FullMethod: "/grpc.testing.TestService/EmptyCall"},
+		&stats.OutPayload{WireLength: 5},
+		&stats.InTrailer{Trailer: metadata.Pairs("content-type", "application/grpc", "grpc-retry-pushback-ms", "10")},
+		&stats.End{},
+
+		&stats.Begin{},
+		&stats.OutHeader{FullMethod: "/grpc.testing.TestService/EmptyCall"},
+		&stats.OutPayload{WireLength: 5},
+		&stats.InHeader{},
+		&stats.InPayload{WireLength: 5},
+		&stats.InTrailer{},
+		&stats.End{},
+	}
+
+	// There is a race between noticing the RST_STREAM during the first RPC
+	// attempt and writing the payload.  If we detect that the client did not
+	// send the OutPayload, we remove it from want.
+	if _, ok := handler.s[2].(*stats.End); ok {
+		want = append(want[:2], want[3:]...)
+	}
+
+	toString := func(ss []stats.RPCStats) (ret []string) {
+		for _, s := range ss {
+			ret = append(ret, fmt.Sprintf("%T - %v", s, s))
+		}
+		return ret
+	}
+	t.Logf("Handler received frames:\n%v\n---\nwant:\n%v\n",
+		strings.Join(toString(handler.s), "\n"),
+		strings.Join(toString(want), "\n"))
+
+	if len(handler.s) != len(want) {
+		t.Fatalf("received unexpected number of RPCStats: got %v; want %v", len(handler.s), len(want))
+	}
+
+	// There is a race between receiving the payload (triggered by the
+	// application / gRPC library) and receiving the trailer (triggered at the
+	// transport layer).  Adjust the received stats accordingly if necessary.
+	// Note: we measure from the end of the RPCStats due to the race above.
+	tIdx, pIdx := len(handler.s)-3, len(handler.s)-2
+	_, okT := handler.s[tIdx].(*stats.InTrailer)
+	_, okP := handler.s[pIdx].(*stats.InPayload)
+	if okT && okP {
+		handler.s[pIdx], handler.s[tIdx] = handler.s[tIdx], handler.s[pIdx]
+	}
+
+	for i := range handler.s {
```

Comment:
Can this also do a `cmp.Diff()`?

`cmp.Comparer(func(a, b time.Time) bool { return true })` to ignore time.Time.

It also has options to only ignore a specific field.

Reply:
I tried this, but didn't think it was worth it.  There are a lot of fields that I don't think it's important to test here (times, some wire lengths, local/remote address, headers, some trailers).

So I either have to exclude a ton of things, including some things conditionally, or set them in the `want` data always.

One thing I noticed that should be tested but wasn't due to my custom comparison logic here was `IsTransparentRetryAttempt=false`.